config/trace.go

```diff
@@ -4,15 +4,174 @@
-		return noop.NewTracerProvider(), noopShutdown
+		return noop.NewTracerProvider(), noopShutdown, nil
+	}
+	opts := []sdktrace.TracerProviderOption{
+		sdktrace.WithResource(res),
+	}
+	for _, processor := range cfg.opentelemetryConfig.TracerProvider.Processors {
+		sp, err := spanProcessor(cfg.ctx, processor)
+		if err != nil {
+			return noop.NewTracerProvider(), noopShutdown, err
+		}
+		opts = append(opts, sdktrace.WithSpanProcessor(sp))
+	}
+	tp := sdktrace.NewTracerProvider(opts...)
+	return tp, tp.Shutdown, nil
+}
+
+func spanExporter(ctx context.Context, exporter SpanExporter) (sdktrace.SpanExporter, error) {
+	if exporter.Console != nil {
+		return stdouttrace.New(
+			stdouttrace.WithPrettyPrint(),
+		)
+	}
+	if exporter.OTLP != nil {
+		switch exporter.OTLP.Protocol {
+		case protocolProtobufHTTP:
+			return initOTLPHTTPSpanExporter(ctx, exporter.OTLP)
+		case protocolProtobufGRPC:
+			return initOTLPgRPCSpanExporter(ctx, exporter.OTLP)
+		default:
+			return nil, fmt.Errorf("unsupported protocol %q", exporter.OTLP.Protocol)
+		}
+	}
+	return nil, errNoValidSpanExporter
+}
+
+func spanProcessor(ctx context.Context, processor SpanProcessor) (sdktrace.SpanProcessor, error) {
+	if processor.Batch != nil {
+		exp, err := spanExporter(ctx, processor.Batch.Exporter)
+		if err != nil {
+			return nil, err
+		}
+		return batchSpanProcessor(processor.Batch, exp)
+	}
+	if processor.Simple != nil {
+		exp, err := spanExporter(ctx, processor.Simple.Exporter)
+		if err != nil {
+			return nil, err
+		}
+		return sdktrace.NewSimpleSpanProcessor(exp), nil
+	}
+	return nil, fmt.Errorf("unsupported span processor type %v", processor)
+}
+
+func initOTLPgRPCSpanExporter(ctx context.Context, otlpConfig *OTLP) (sdktrace.SpanExporter, error) {
+	opts := []otlptracegrpc.Option{}
+
+	if len(otlpConfig.Endpoint) > 0 {
+		u, err := url.ParseRequestURI(normalizeEndpoint(otlpConfig.Endpoint))
+		if err != nil {
+			return nil, err
+		}
+		opts = append(opts, otlptracegrpc.WithEndpoint(u.Host))
+		if u.Scheme == "http" {
+			opts = append(opts, otlptracegrpc.WithInsecure())
+		}
+	}
+
+	if otlpConfig.Compression != nil {
+		switch *otlpConfig.Compression {
+		case "gzip":
+			opts = append(opts, otlptracegrpc.WithCompressor(*otlpConfig.Compression))
+		case "none":
+			// none requires no options
+		default:
+			return nil, fmt.Errorf("unsupported compression %q", *otlpConfig.Compression)
+		}
+	}
+	if otlpConfig.Timeout != nil && *otlpConfig.Timeout > 0 {
+		opts = append(opts, otlptracegrpc.WithTimeout(time.Millisecond*time.Duration(*otlpConfig.Timeout)))
+	}
+	if len(otlpConfig.Headers) > 0 {
+		opts = append(opts, otlptracegrpc.WithHeaders(otlpConfig.Headers))
+	}
+
+	return otlptracegrpc.New(ctx, opts...)
+}
+
+func initOTLPHTTPSpanExporter(ctx context.Context, otlpConfig *OTLP) (sdktrace.SpanExporter, error) {
+	opts := []otlptracehttp.Option{}
+
+	if len(otlpConfig.Endpoint) > 0 {
+		u, err := url.ParseRequestURI(normalizeEndpoint(otlpConfig.Endpoint))
```

Comment:
`normalizeEndpoint` should not be used here.

It would break https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#endpoint-urls-for-otlphttp:

> An SDK MUST NOT modify the URL in ways other than specified above.

Reply:
@codeboten What about this comment?

Reply:
yes I haven't addressed it yet. will do so today

Reply:
@pellared removed the function

Reply:
Sorry, but I my comment was not very explicit. The function should not be used in **this** place. I think it is required for parsing the endpoint for gRPC protocol 😬

Reply:
I ended up doing updating the grpc case to try parsing the URL, and if the scheme isn't present (which helpfully `ParseRequestURI` puts the host in the place of the scheme 😆) it will use the full endpoint string instead of u.Host.

So for example if endpoint is `http://localhost:4317`, the following options are set:

```
WithEndpoint("localhost:4317")
WithInsecure()
```

If the endpoint string is `localhost:4317`, the following option is set:

```
WithEndpoint("localhost:4317")
```

I think this makes sense, because someone may want to set `localhost:4317` and have it be a secure option. WDYT?

Reply:
I see. Can you just add comments to the gRPC parsing case as this is hacky and it will be hard to understand in future?